{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Example usage of GarNet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# if you pip installed the package: \n",
    "import GarNet\n",
    "\n",
    "# if you cloned the repo:\n",
    "# import sys\n",
    "# sys.path.append(\"../GarNet\")\n",
    "\n",
    "# Having issues importing despite module exising in virtual environment? \n",
    "# See https://github.com/jupyter/notebook/issues/397"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Generate GarNet file\n",
    "\n",
    "Skip this section if GarNet file has already been generated. See Appendix for how to obtain UCSC reference file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "04:36:41 - GarNet: INFO -   - TSS file already exists here ../garnet_data/reference/ucsc_reference.hg19.tss.bed\n",
      "04:36:41 - GarNet: INFO -   - Searching for motifs near genes. This may take a while...\n",
      "04:36:48 - GarNet: INFO -   - 181422 motif-gene associations found and written to ../garnet_data/cisBP_hg19_SLIM/garnetDB_cisBP.hg19.LOD.10kb.chr1_SLIM.tsv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>motifChrom</th>\n",
       "      <th>motifStart</th>\n",
       "      <th>motifEnd</th>\n",
       "      <th>motifName</th>\n",
       "      <th>motifScore</th>\n",
       "      <th>motifStrand</th>\n",
       "      <th>geneName</th>\n",
       "      <th>tssStart</th>\n",
       "      <th>tssEnd</th>\n",
       "      <th>motif_gene_distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17000</td>\n",
       "      <td>17008</td>\n",
       "      <td>HOXA4</td>\n",
       "      <td>8.846505</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>11873</td>\n",
       "      <td>11874</td>\n",
       "      <td>5135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>18055</td>\n",
       "      <td>18067</td>\n",
       "      <td>LMO2</td>\n",
       "      <td>8.768321</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>11873</td>\n",
       "      <td>11874</td>\n",
       "      <td>6194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>18339</td>\n",
       "      <td>18351</td>\n",
       "      <td>LMO2</td>\n",
       "      <td>8.901481</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>11873</td>\n",
       "      <td>11874</td>\n",
       "      <td>6478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>19086</td>\n",
       "      <td>19096</td>\n",
       "      <td>SREBF1</td>\n",
       "      <td>11.043803</td>\n",
       "      <td>-</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>11873</td>\n",
       "      <td>11874</td>\n",
       "      <td>7213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>19858</td>\n",
       "      <td>19866</td>\n",
       "      <td>HOXA4</td>\n",
       "      <td>8.539458</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>11873</td>\n",
       "      <td>11874</td>\n",
       "      <td>7993</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  motifChrom  motifStart  motifEnd motifName  motifScore motifStrand geneName  \\\n",
       "0       chr1       17000     17008     HOXA4    8.846505           +  DDX11L1   \n",
       "1       chr1       18055     18067      LMO2    8.768321           +  DDX11L1   \n",
       "2       chr1       18339     18351      LMO2    8.901481           +  DDX11L1   \n",
       "3       chr1       19086     19096    SREBF1   11.043803           -  DDX11L1   \n",
       "4       chr1       19858     19866     HOXA4    8.539458           +  DDX11L1   \n",
       "\n",
       "   tssStart  tssEnd  motif_gene_distance  \n",
       "0     11873   11874                 5135  \n",
       "1     11873   11874                 6194  \n",
       "2     11873   11874                 6478  \n",
       "3     11873   11874                 7213  \n",
       "4     11873   11874                 7993  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reference_filepath = \"../garnet_data/reference/ucsc_reference.hg19.bed\"\n",
    "motif_filepath = \"../garnet_data/cisBP_hg19_SLIM/cisBP.hg19.LOD.10kb.chr1_SLIM.bed\"\n",
    "garnet_filepath = \"../garnet_data/cisBP_hg19_SLIM/garnetDB_cisBP.hg19.LOD.10kb.chr1_SLIM.tsv\"\n",
    "\n",
    "garnet_df = GarNet.construct_garnet_file(reference_filepath, motif_filepath, garnet_filepath)\n",
    "garnet_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Map peaks to genes using a549 data\n",
    "\n",
    "This is accomplished by intersecting motifs with genes specified in the open chromatin bed file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>motifName</th>\n",
       "      <th>motifScore</th>\n",
       "      <th>motifStrand</th>\n",
       "      <th>geneName</th>\n",
       "      <th>geneStart</th>\n",
       "      <th>geneEnd</th>\n",
       "      <th>motif_gene_distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>8086390</td>\n",
       "      <td>8086402</td>\n",
       "      <td>AC0021266</td>\n",
       "      <td>10.431478</td>\n",
       "      <td>-</td>\n",
       "      <td>ERRFI1</td>\n",
       "      <td>8086392</td>\n",
       "      <td>8086393</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>23913349</td>\n",
       "      <td>23913361</td>\n",
       "      <td>AC0021266</td>\n",
       "      <td>13.024280</td>\n",
       "      <td>-</td>\n",
       "      <td>MDS2</td>\n",
       "      <td>23907984</td>\n",
       "      <td>23907985</td>\n",
       "      <td>5365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24246686</td>\n",
       "      <td>24246698</td>\n",
       "      <td>AC0021266</td>\n",
       "      <td>12.883207</td>\n",
       "      <td>+</td>\n",
       "      <td>CNR2</td>\n",
       "      <td>24239816</td>\n",
       "      <td>24239817</td>\n",
       "      <td>-6882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24246686</td>\n",
       "      <td>24246698</td>\n",
       "      <td>AC0021266</td>\n",
       "      <td>12.883207</td>\n",
       "      <td>+</td>\n",
       "      <td>MIR378F</td>\n",
       "      <td>24255559</td>\n",
       "      <td>24255560</td>\n",
       "      <td>-8861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>57297772</td>\n",
       "      <td>57297784</td>\n",
       "      <td>AC0021266</td>\n",
       "      <td>12.783137</td>\n",
       "      <td>-</td>\n",
       "      <td>BC048114</td>\n",
       "      <td>57292592</td>\n",
       "      <td>57292593</td>\n",
       "      <td>-5180</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom     start       end  motifName  motifScore motifStrand  geneName  \\\n",
       "0  chr1   8086390   8086402  AC0021266   10.431478           -    ERRFI1   \n",
       "1  chr1  23913349  23913361  AC0021266   13.024280           -      MDS2   \n",
       "2  chr1  24246686  24246698  AC0021266   12.883207           +      CNR2   \n",
       "3  chr1  24246686  24246698  AC0021266   12.883207           +   MIR378F   \n",
       "4  chr1  57297772  57297784  AC0021266   12.783137           -  BC048114   \n",
       "\n",
       "   geneStart   geneEnd  motif_gene_distance  \n",
       "0    8086392   8086393                    2  \n",
       "1   23907984  23907985                 5365  \n",
       "2   24239816  24239817                -6882  \n",
       "3   24255559  24255560                -8861  \n",
       "4   57292592  57292593                -5180  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "garnet_filepath = \"../data/garnetDB_cisBP.hg19.LOD.10kb.tsv\"\n",
    "peaks_filepath = \"a549_example_peaks.chr1.bed\"\n",
    "\n",
    "df = GarNet.map_peaks(peaks_filepath, garnet_filepath)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Regress gene expression against TF Motif binding score"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Perform regression, plots are automatically generated in `regression_plots/`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "04:48:00 - GarNet: INFO - Performing linear regression for 692 transcription factor expression profiles...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Transcription Factor</th>\n",
       "      <th>Slope</th>\n",
       "      <th>P-Value</th>\n",
       "      <th>Targets</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>FOXO1</td>\n",
       "      <td>1.486838</td>\n",
       "      <td>0.004396</td>\n",
       "      <td>PVRL4,TMEM63A,S100A14,EFNA1,SYTL1,LRRC8C,TGFB2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>FOXO6</td>\n",
       "      <td>1.199703</td>\n",
       "      <td>0.008159</td>\n",
       "      <td>GALNT2,SYTL1,PVRL4,TMEM63A,SOX13,ELF3,LAD1,S10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>FOXL1</td>\n",
       "      <td>-2.024675</td>\n",
       "      <td>0.008195</td>\n",
       "      <td>TMEM63A,SOX13,ELF3,PVRL4,EFNA1,CYR61,LRRC8C,PT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>TEAD3</td>\n",
       "      <td>0.352063</td>\n",
       "      <td>0.011501</td>\n",
       "      <td>PTGS2,SOX13,BCAN,NMNAT2,ZBTB7B,ACOT7,ADAM15,LA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>VDR</td>\n",
       "      <td>0.968223</td>\n",
       "      <td>0.012346</td>\n",
       "      <td>KIRREL,RAB25,SYTL1,CSF1,ELF3,PTPN14,TMEM61,LAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366</th>\n",
       "      <td>ZNF274</td>\n",
       "      <td>0.818921</td>\n",
       "      <td>0.014866</td>\n",
       "      <td>LEPRE1,NMNAT2,PVRL4,LRRC8B,S100A14,MFAP2,DVL1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>RXRA</td>\n",
       "      <td>-1.486082</td>\n",
       "      <td>0.014930</td>\n",
       "      <td>ANXA9,RAB25,S100A14,ELF3,ARHGEF16,MFAP2,LMNA,A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>NFIC</td>\n",
       "      <td>-0.074088</td>\n",
       "      <td>0.025818</td>\n",
       "      <td>KIAA1522,ZBTB7B,ATF3,SOX13,BCAN,TACSTD2,THBS3,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>RFX1</td>\n",
       "      <td>0.832647</td>\n",
       "      <td>0.033029</td>\n",
       "      <td>CAMK2N1,ZBTB7B,PVRL4,LMNA,RPS6KA1,THBS3,LIX1L,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>HIC2</td>\n",
       "      <td>-15.151852</td>\n",
       "      <td>0.035823</td>\n",
       "      <td>S100A4,RPS6KA1,ELF3,ANXA9,PTAFR,LMNA,YARS,S100A6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Transcription Factor      Slope   P-Value  \\\n",
       "94                 FOXO1   1.486838  0.004396   \n",
       "97                 FOXO6   1.199703  0.008159   \n",
       "89                 FOXL1  -2.024675  0.008195   \n",
       "315                TEAD3   0.352063  0.011501   \n",
       "341                  VDR   0.968223  0.012346   \n",
       "366               ZNF274   0.818921  0.014866   \n",
       "253                 RXRA  -1.486082  0.014930   \n",
       "193                 NFIC  -0.074088  0.025818   \n",
       "240                 RFX1   0.832647  0.033029   \n",
       "115                 HIC2 -15.151852  0.035823   \n",
       "\n",
       "                                               Targets  \n",
       "94   PVRL4,TMEM63A,S100A14,EFNA1,SYTL1,LRRC8C,TGFB2...  \n",
       "97   GALNT2,SYTL1,PVRL4,TMEM63A,SOX13,ELF3,LAD1,S10...  \n",
       "89   TMEM63A,SOX13,ELF3,PVRL4,EFNA1,CYR61,LRRC8C,PT...  \n",
       "315  PTGS2,SOX13,BCAN,NMNAT2,ZBTB7B,ACOT7,ADAM15,LA...  \n",
       "341  KIRREL,RAB25,SYTL1,CSF1,ELF3,PTPN14,TMEM61,LAD...  \n",
       "366  LEPRE1,NMNAT2,PVRL4,LRRC8B,S100A14,MFAP2,DVL1,...  \n",
       "253  ANXA9,RAB25,S100A14,ELF3,ARHGEF16,MFAP2,LMNA,A...  \n",
       "193  KIAA1522,ZBTB7B,ATF3,SOX13,BCAN,TACSTD2,THBS3,...  \n",
       "240  CAMK2N1,ZBTB7B,PVRL4,LMNA,RPS6KA1,THBS3,LIX1L,...  \n",
       "115   S100A4,RPS6KA1,ELF3,ANXA9,PTAFR,LMNA,YARS,S100A6  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_filepath = \"a549_example_expression.txt\"\n",
    "\n",
    "results = GarNet.TF_regression(df, expression_filepath, \".\")\n",
    "results.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Appendix\n",
    "\n",
    "## Download and format reference file\n",
    "\n",
    "### 1. Download UCSC known genes file\n",
    "\n",
    "Visit [UCSC Table Browser](https://genome.ucsc.edu/cgi-bin/hgTables?command=start) and select the following options\n",
    "\n",
    " - **assembly**: Feb. 2009 (GRCh37/hg19)\n",
    " - **group**: Genes and Gene Predictions\n",
    " - **track**: UCSC Genes\n",
    " - **table**: knownGene\n",
    " - **region**: genome\n",
    " - **output format**: selected fields from primary and related tables\n",
    "\n",
    "Press \"get output\", clear all fields, then check the following fields:\n",
    "\n",
    " - chrom\n",
    " - strand\n",
    " - txStart\n",
    " - txEnd\n",
    " - geneSymbol\n",
    " \n",
    "Press \"get output\" again, and save the table locally. A copy (retrieved 9/11/2017) is saved in `../garnet_data/reference/ucsc_known_genes.hg19.tsv`.\n",
    "\n",
    "### 2. Reformat to BED file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>11873</td>\n",
       "      <td>14409</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>14361</td>\n",
       "      <td>16765</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>16857</td>\n",
       "      <td>17751</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>15795</td>\n",
       "      <td>18061</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>14361</td>\n",
       "      <td>19759</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom strand  chromStart  chromEnd     name score\n",
       "0  chr1      +       11873     14409  DDX11L1     .\n",
       "3  chr1      -       14361     16765   WASH7P     .\n",
       "4  chr1      -       16857     17751   WASH7P     .\n",
       "5  chr1      -       15795     18061   WASH7P     .\n",
       "6  chr1      -       14361     19759   WASH7P     ."
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ucsc_reference = pd.read_csv(\"../garnet_data/reference/ucsc_known_genes.hg19.tsv\", sep='\\t', names=[\"chrom\", \"strand\", \"chromStart\", \"chromEnd\", \"name\"], comment=\"#\")\n",
    "ucsc_reference.drop_duplicates(inplace=True)\n",
    "ucsc_reference[\"score\"] = \".\"\n",
    "ucsc_reference.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11873</td>\n",
       "      <td>14409</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14361</td>\n",
       "      <td>16765</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16857</td>\n",
       "      <td>17751</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15795</td>\n",
       "      <td>18061</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14361</td>\n",
       "      <td>19759</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  chromStart  chromEnd     name score strand\n",
       "0  chr1       11873     14409  DDX11L1     .      +\n",
       "3  chr1       14361     16765   WASH7P     .      -\n",
       "4  chr1       16857     17751   WASH7P     .      -\n",
       "5  chr1       15795     18061   WASH7P     .      -\n",
       "6  chr1       14361     19759   WASH7P     .      -"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ucsc_reference_bed = ucsc_reference[[\"chrom\", \"chromStart\", \"chromEnd\", \"name\", \"score\", \"strand\"]]\n",
    "ucsc_reference_bed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "ucsc_reference_bed.to_csv(\"../garnet_data/reference/ucsc_reference.hg19.bed\", sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
