<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8">
    
    <title>Welcome to GarNet’s technical documentation! &mdash; GarNet 2.0 documentation</title>
    
    <link rel="stylesheet" type="text/css" href="_static/css/spc-bootstrap.css">
    <link rel="stylesheet" type="text/css" href="_static/css/spc-extend.css">
    <link rel="stylesheet" href="_static/scipy.css" type="text/css" >
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" >
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '2.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="_static/js/copybutton.js"></script>
    <link rel="index" title="Index" href="genindex.html" >
    <link rel="search" title="Search" href="search.html" >
    <link rel="top" title="GarNet 2.0 documentation" href="#" > 
  </head>
  <body>

  <div class="container">
    <div class="header">
    </div>
  </div>


    <div class="container">
      <div class="main">
        

	<div class="row-fluid"><div class="span1"></div>
          <div class="span10">
	  
            
	      <div class="row-fluid">
		<div class="span9">
  		  <div class="spc-navbar">
		    
    <ul class="nav nav-pills pull-left">
        <li class="active"><a href="["></a></li>
        <li class="active"><a href="]"></a></li>
	 
    </ul>
		  </div>
		</div>
	      </div>
            
        <div class="bodywrapper">
          <div class="body" id="spc-section-body">
            
  <div class="section" id="welcome-to-garnet-s-technical-documentation">
<h1>Welcome to GarNet&#8217;s technical documentation!<a class="headerlink" href="#welcome-to-garnet-s-technical-documentation" title="Permalink to this headline">¶</a></h1>
<p>GarNet is an epigenomics data exploration tool. For more information about the scientific uses of GarNet,
please see the <a class="reference external" href="https://github.com/fraenkel-lab/GarNet">Readme</a>. To report an error, please refer to the <a class="reference external" href="https://github.com/fraenkel-lab/GarNet/issues">Issues</a>.</p>
<p>Some important notes:</p>
<ul class="simple">
<li>GarNet uses tab-delimited files, otherwise known as tsv&#8217;s. BED files are already tsv, so no worries there.</li>
<li>GarNet saves intermediate files for all input datasets, which it can use instead of raw tsv&#8217;s saving computation and time. Those are pickled IntervalTrees (files which end in .pickle)</li>
<li>We&#8217;ve supplied such intermediate files for motifs and genes in the repo, leaving it to you to supply the peaks file and the expression file.</li>
<li>GarNet uses an Options class for all parameters. This little snipped gives you a dict whose members you can access with . instead of []</li>
</ul>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">Options</span><span class="p">:</span>
        <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">options</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="vm">__dict__</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">options</span><span class="p">)</span>
</pre></div>
</div>
<ul class="simple">
<li>The options mentioned in the following documentation are expected as Options objects, with the fields mentioned included.</li>
</ul>
<span class="target" id="module-garnet"></span><dl class="function">
<dt id="garnet.map_known_genes_and_motifs_to_peaks">
<code class="descclassname">garnet.</code><code class="descname">map_known_genes_and_motifs_to_peaks</code><span class="sig-paren">(</span><em>known_genes_file</em>, <em>motifs_file</em>, <em>peaks_file</em>, <em>options</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/garnet.html#map_known_genes_and_motifs_to_peaks"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#garnet.map_known_genes_and_motifs_to_peaks" title="Permalink to this definition">¶</a></dt>
<dd><p>Find motifs and associated genes local to peaks.</p>
<p>This function searches for motifs &#8220;under&#8221; peaks from an epigenomics dataset and &#8220;around&#8221; peaks for genes.
It then returns all pairs of motifs and genes which were found local to peaks.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>peaks_file</strong> (<em>str</em><em> or </em><em>FILE</em>) &#8211; filepath or file object for the peaks file.</li>
<li><strong>known_genes_file</strong> (<em>str</em><em> or </em><em>FILE</em>) &#8211; filepath or file object for the known_genes file</li>
<li><strong>motifs_file</strong> (<em>str</em><em> or </em><em>FILE</em>) &#8211; filepath or file object for the motifs file</li>
<li><strong>options</strong> (<em>Options</em>) &#8211; Options({&#8220;upstream_window&#8221;: int, &#8220;downstream_window&#8221;: int, &#8220;tss&#8221;: bool, &#8220;output_dir&#8221;: string (optional)})</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><dl class="docutils">
<dt>a dataframe with rows of transcription factor binding motifs and nearby genes</dt>
<dd><p class="first last">with the restriction that these motifs and genes must have been found near a peak.</p>
</dd>
</dl>
</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">dataframe</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="garnet.map_known_genes_to_peaks">
<code class="descclassname">garnet.</code><code class="descname">map_known_genes_to_peaks</code><span class="sig-paren">(</span><em>known_genes_file</em>, <em>peaks_file</em>, <em>options</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/garnet.html#map_known_genes_to_peaks"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#garnet.map_known_genes_to_peaks" title="Permalink to this definition">¶</a></dt>
<dd><p>Find all genes nearby to peaks.</p>
<p>This function searches in the neighborhood of peaks for genes and returns each peak / gene pair
which were found to be local to one another.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>peaks_file</strong> (<em>str</em><em> or </em><em>FILE</em>) &#8211; filepath or file object for the peaks file.</li>
<li><strong>known_genes_file</strong> (<em>str</em><em> or </em><em>FILE</em>) &#8211; filepath or file object for the known_genes file</li>
<li><strong>options</strong> (<em>Options</em>) &#8211; Options({&#8220;upstream_window&#8221;: int, &#8220;downstream_window&#8221;: int, &#8220;tss&#8221;: bool, &#8220;output_dir&#8221;: string (optional)})</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">A dataframe listing peaks and nearby genes</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">dataframe</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="garnet.map_motifs_to_peaks">
<code class="descclassname">garnet.</code><code class="descname">map_motifs_to_peaks</code><span class="sig-paren">(</span><em>motifs_file</em>, <em>peaks_file</em>, <em>options</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/garnet.html#map_motifs_to_peaks"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#garnet.map_motifs_to_peaks" title="Permalink to this definition">¶</a></dt>
<dd><p>Find known transcription factor binding motifs motifs &#8220;below&#8221; input epigenetic peaks.</p>
<p>This function searches for overlap of motifs and peaks, and returns peak / motif pairs.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>peaks_file</strong> (<em>str</em><em> or </em><em>FILE</em>) &#8211; filepath or file object for the peaks file.</li>
<li><strong>motifs_file</strong> (<em>str</em><em> or </em><em>FILE</em>) &#8211; filepath or file object for the motifs file</li>
<li><strong>options</strong> (<em>Options</em>) &#8211; Options({&#8220;output_dir&#8221;: string (optional)})</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">A dataframe listing peaks and nearby transcription factor binding motifs</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">dataframe</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="garnet.map_known_genes_to_motifs">
<code class="descclassname">garnet.</code><code class="descname">map_known_genes_to_motifs</code><span class="sig-paren">(</span><em>known_genes_file</em>, <em>motifs_file</em>, <em>options</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/garnet.html#map_known_genes_to_motifs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#garnet.map_known_genes_to_motifs" title="Permalink to this definition">¶</a></dt>
<dd><p>Associate genes local to motifs with those motifs, without peak information.</p>
<p>This function searches for overlap of motifs and genes, and returns motif / gene pairs.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>known_genes_file</strong> (<em>str</em><em> or </em><em>FILE</em>) &#8211; filepath or file object for the known_genes file</li>
<li><strong>motifs_file</strong> (<em>str</em><em> or </em><em>FILE</em>) &#8211; filepath or file object for the motifs file</li>
<li><strong>options</strong> (<em>Options</em>) &#8211; Options({&#8220;upstream_window&#8221;: int, &#8220;downstream_window&#8221;: int, &#8220;tss&#8221;: bool, &#8220;output_dir&#8221;: string (optional)})</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">A dataframe listing transcription factor binding motifs and nearby genes.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">dataframe</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="garnet.TF_regression">
<code class="descclassname">garnet.</code><code class="descname">TF_regression</code><span class="sig-paren">(</span><em>motifs_and_genes_dataframe</em>, <em>expression_file</em>, <em>options</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/garnet.html#TF_regression"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#garnet.TF_regression" title="Permalink to this definition">¶</a></dt>
<dd><p>Do linear regression of the expression of genes versus the strength of the assiciated transcription factor binding motifs and report results.</p>
<p>This function parses an expression file of two columns: gene symbol and expression value, and
merges the expression profile into the motifs and genes file, resulting in information about
transcription factor binding motifs local to genes, and those genes&#8217; expressions. We do linear
regression, and if an output directory is provided, we output a plot for each TF and an html
summary of the regressions.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>motifs_and_genes_dataframe</strong> (<em>dataframe</em>) &#8211; the outcome of map_known_genes_and_motifs_to_peaks</li>
<li><strong>expression_file</strong> (<em>str</em><em> or </em><em>FILE</em>) &#8211; a tsv file of expression data, with geneSymbol, score columns</li>
<li><strong>options</strong> (<em>Options</em>) &#8211; Options({&#8220;output_dir&#8221;: string (optional)})</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">slope and pval of linear regfression for each transcription factor.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">dataframe</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>


          </div>
        </div>
          </div><div class="span1"></div>
        </div>
      </div>
    </div>

    <div class="container container-navbar-bottom">
      <div class="spc-navbar">
        
      </div>
    </div>
    <div class="container">
    <div class="footer">
    <div class="row-fluid">
    <ul class="inline pull-left">
      <li>
        &copy; Copyright 2017, Alexander LeNail.
      </li>
      <li>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.5.2.
      </li>
    </ul>
    </div>
    </div>
    </div>
  </body>
</html>