

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>garnet &mdash; GarNet 2.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../genindex.html"/>
        <link rel="search" title="Search" href="../search.html"/>
    <link rel="top" title="GarNet 2.0 documentation" href="../index.html"/>
        <link rel="up" title="Module code" href="index.html"/> 

  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> GarNet
          

          
          </a>

          
            
            
              <div class="version">
                2.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <!-- Local TOC -->
              <div class="local-toc"></div>
            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">GarNet</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="index.html">Module code</a> &raquo;</li>
        
      <li>garnet</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for garnet</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python3</span>

<span class="c1"># Core python modules</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">os</span>

<span class="c1"># Peripheral python modules</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">subprocess</span>

<span class="c1"># Core python external libraries</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">statsmodels.formula.api</span> <span class="k">import</span> <span class="n">ols</span> <span class="k">as</span> <span class="n">linear_regression</span>
<span class="kn">from</span> <span class="nn">statsmodels.graphics.regressionplots</span> <span class="k">import</span> <span class="n">abline_plot</span> <span class="k">as</span> <span class="n">plot_regression</span>

<span class="c1"># Peripheral python external libraries</span>
<span class="kn">from</span> <span class="nn">pybedtools</span> <span class="k">import</span> <span class="n">BedTool</span>
<span class="kn">import</span> <span class="nn">jinja2</span>

<span class="c1"># list of public methods:</span>
<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span> <span class="s2">&quot;construct_garnet_file&quot;</span><span class="p">,</span> <span class="s2">&quot;map_peaks&quot;</span><span class="p">,</span> <span class="s2">&quot;TF_regression&quot;</span> <span class="p">]</span>

<span class="n">templateLoader</span> <span class="o">=</span> <span class="n">jinja2</span><span class="o">.</span><span class="n">FileSystemLoader</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="vm">__file__</span><span class="p">)))</span>
<span class="n">templateEnv</span> <span class="o">=</span> <span class="n">jinja2</span><span class="o">.</span><span class="n">Environment</span><span class="p">(</span><span class="n">loader</span><span class="o">=</span><span class="n">templateLoader</span><span class="p">)</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>
<span class="n">logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>
<span class="n">handler</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">StreamHandler</span><span class="p">()</span>
<span class="n">handler</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>
<span class="n">handler</span><span class="o">.</span><span class="n">setFormatter</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">Formatter</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%(asctime)s</span><span class="s1"> - GarNet: </span><span class="si">%(levelname)s</span><span class="s1"> - </span><span class="si">%(message)s</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s2">&quot;%I:%M:%S&quot;</span><span class="p">))</span>
<span class="n">logger</span><span class="o">.</span><span class="n">addHandler</span><span class="p">(</span><span class="n">handler</span><span class="p">)</span>


<span class="c1">######################################## File Parsing Logic #######################################</span>

<span class="k">def</span> <span class="nf">parse_expression_file</span><span class="p">(</span><span class="n">expression_file</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Parse gene expression scores from a transcriptomics assay (e.g. RNAseq) into a dataframe and ensures</span>
<span class="sd">	columns are of the correct type.</span>

<span class="sd">	Arguments:</span>
<span class="sd">		expression_file (string or FILE): Two-column, tab-delimited file of gene / gene expression score</span>

<span class="sd">	Returns:</span>
<span class="sd">		pd.dataframe: expression dataframe</span>
<span class="sd">	&quot;&quot;&quot;</span>
	<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">expression_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;expression&quot;</span><span class="p">])</span>
	<span class="n">df</span><span class="p">[</span><span class="s2">&quot;expression&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">to_numeric</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;expression&quot;</span><span class="p">],</span> <span class="n">errors</span><span class="o">=</span><span class="s2">&quot;coerce&quot;</span><span class="p">)</span>
	<span class="n">df</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

	<span class="k">return</span> <span class="n">df</span>


<span class="k">def</span> <span class="nf">_parse_motifs_and_genes_file_or_dataframe</span><span class="p">(</span><span class="n">motifs_and_genes_file_or_dataframe</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	If the argument is a dataframe, return it. Otherwise if the argument is a string, try to read a dataframe from it, and return that</span>
<span class="sd">	&quot;&quot;&quot;</span>

	<span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">motifs_and_genes_file_or_dataframe</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
		<span class="n">motifs_and_genes_dataframe</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">motifs_and_genes_file_or_dataframe</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

	<span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">motifs_and_genes_file_or_dataframe</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
		<span class="n">motifs_and_genes_dataframe</span> <span class="o">=</span> <span class="n">motifs_and_genes_file_or_dataframe</span>

	<span class="k">else</span><span class="p">:</span> <span class="n">logger</span><span class="o">.</span><span class="n">critical</span><span class="p">(</span><span class="s1">&#39;argument not recognized as a file or a dataframe, exiting...&#39;</span><span class="p">);</span> <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

	<span class="k">return</span> <span class="n">motifs_and_genes_dataframe</span>


<span class="c1">######################################### Public Functions #########################################</span>

<div class="viewcode-block" id="map_peaks"><a class="viewcode-back" href="../index.html#garnet.map_peaks">[docs]</a><span class="k">def</span> <span class="nf">map_peaks</span><span class="p">(</span><span class="n">peaks_filepath_or_list_of_peaks_filepaths</span><span class="p">,</span> <span class="n">garnet_filepath</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Find motifs and associated genes local to peaks.</span>

<span class="sd">	This function intersects peaks from an epigenomics dataset with TF motifs.</span>

<span class="sd">	Arguments:</span>
<span class="sd">		peaks_filepath_or_list_of_peaks_filepaths (str or list): filepath of the peaks file, or list of such paths</span>
<span class="sd">		garnet_filepath (str): filepath to the garnet file.</span>

<span class="sd">	Returns:</span>
<span class="sd">		pd.DataFrame: a dataframe with rows of transcription factor binding motifs and nearby genes with</span>
<span class="sd">		the restriction that these motifs and genes must have been found near a peak.</span>
<span class="sd">	&quot;&quot;&quot;</span>

	<span class="c1"># peaks_filepath_or_list_of_peaks_filepaths is either a filepath or FILE, or a list of filepaths or FILEs.</span>
	<span class="c1"># Let&#39;s operate on a list in either case, so if it&#39;s a single string, put it in a list. #TODO, this will break if it&#39;s a single FILE.</span>
	<span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">peaks_filepath_or_list_of_peaks_filepaths</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span> <span class="n">list_of_peaks_filepaths</span> <span class="o">=</span> <span class="n">peaks_filepath_or_list_of_peaks_filepaths</span>
	<span class="k">else</span><span class="p">:</span> <span class="n">list_of_peaks_filepaths</span> <span class="o">=</span> <span class="p">[</span><span class="n">peaks_filepath_or_list_of_peaks_filepaths</span><span class="p">]</span>
	<span class="k">assert</span> <span class="nb">all</span><span class="p">([</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">peaks_filepath</span><span class="p">)</span> <span class="k">for</span> <span class="n">peaks_filepath</span> <span class="ow">in</span> <span class="n">list_of_peaks_filepaths</span><span class="p">])</span>

	<span class="n">output</span> <span class="o">=</span> <span class="p">[]</span>

	<span class="k">for</span> <span class="n">peaks_filepath</span> <span class="ow">in</span> <span class="n">list_of_peaks_filepaths</span><span class="p">:</span>

		<span class="n">peaks</span> <span class="o">=</span> <span class="n">BedTool</span><span class="p">(</span><span class="n">peaks_filepath</span><span class="p">)</span>
		<span class="n">motifs</span> <span class="o">=</span> <span class="n">BedTool</span><span class="p">(</span><span class="n">garnet_filepath</span><span class="p">)</span>

		<span class="n">intersected</span> <span class="o">=</span> <span class="n">motifs</span><span class="o">.</span><span class="n">intersect</span><span class="p">(</span><span class="n">peaks</span><span class="p">,</span> <span class="n">wa</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">f</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
		<span class="n">intersected_df</span> <span class="o">=</span> <span class="n">intersected</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">(</span><span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">,</span> <span class="s2">&quot;motifName&quot;</span><span class="p">,</span> <span class="s2">&quot;motifScore&quot;</span><span class="p">,</span> <span class="s2">&quot;motifStrand&quot;</span><span class="p">,</span> <span class="s2">&quot;geneName&quot;</span><span class="p">,</span> <span class="s2">&quot;geneStart&quot;</span><span class="p">,</span> <span class="s2">&quot;geneEnd&quot;</span><span class="p">,</span> <span class="s2">&quot;motif_gene_distance&quot;</span><span class="p">])</span>

		<span class="n">output</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">intersected_df</span><span class="p">)</span>

	<span class="c1"># If this function was passed a single file, return a single dataframe</span>
	<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">output</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span> <span class="n">output</span> <span class="o">=</span> <span class="n">output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
	<span class="k">return</span> <span class="n">output</span></div>


<div class="viewcode-block" id="TF_regression"><a class="viewcode-back" href="../index.html#garnet.TF_regression">[docs]</a><span class="k">def</span> <span class="nf">TF_regression</span><span class="p">(</span><span class="n">motifs_and_genes_file_or_dataframe</span><span class="p">,</span> <span class="n">expression_file</span><span class="p">,</span> <span class="n">output_dir</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Do linear regression of the expression of genes versus the strength of the assiciated transcription factor binding motifs and report results.</span>

<span class="sd">	This function parses an expression file of two columns: gene symbol and expression value, and</span>
<span class="sd">	merges the expression profile into the motifs and genes file, resulting in information about</span>
<span class="sd">	transcription factor binding motifs local to genes, and those genes&#39; expressions. We do linear</span>
<span class="sd">	regression, and if an output directory is provided, we output a plot for each TF and an html</span>
<span class="sd">	summary of the regressions.</span>

<span class="sd">	Arguments:</span>
<span class="sd">		motifs_and_genes_file_or_dataframe (str or dataframe): the outcome of map_known_genes_and_motifs_to_peaks, either as a dataframe or a file</span>
<span class="sd">		expression_file (str or FILE): a tsv file of expression data, with geneName, score columns</span>
<span class="sd">		output_dir: (str): If you would like to output figures and a summary html page, supply an output directory</span>

<span class="sd">	Returns:</span>
<span class="sd">		pd.dataframe: slope, pval, and gene targets for each transcription factor.</span>
<span class="sd">	&quot;&quot;&quot;</span>

	<span class="n">motifs_and_genes_dataframe</span> <span class="o">=</span> <span class="n">_parse_motifs_and_genes_file_or_dataframe</span><span class="p">(</span><span class="n">motifs_and_genes_file_or_dataframe</span><span class="p">)</span>
	<span class="n">expression_dataframe</span> <span class="o">=</span> <span class="n">parse_expression_file</span><span class="p">(</span><span class="n">expression_file</span><span class="p">)</span>

	<span class="n">motifs_genes_and_expression_levels</span> <span class="o">=</span> <span class="n">motifs_and_genes_dataframe</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">expression_dataframe</span><span class="p">,</span> <span class="n">left_on</span><span class="o">=</span><span class="s1">&#39;geneName&#39;</span><span class="p">,</span> <span class="n">right_on</span><span class="o">=</span><span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">)</span>

	<span class="c1"># the same geneName might have different names but since the expression is geneName-wise</span>
	<span class="c1"># keep the closest motif to gene in the case of duplicates</span>
	<span class="c1"># TODO: implement function to combine duplicates.</span>
	<span class="k">if</span> <span class="s1">&#39;geneName&#39;</span> <span class="ow">in</span> <span class="n">motifs_genes_and_expression_levels</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
		<span class="n">motifs_genes_and_expression_levels</span><span class="p">[</span><span class="s2">&quot;abs_distance&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">motifs_genes_and_expression_levels</span><span class="o">.</span><span class="n">motif_gene_distance</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span>
		<span class="n">motifs_genes_and_expression_levels</span> <span class="o">=</span> <span class="n">motifs_genes_and_expression_levels</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;motifScore&quot;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> \
		                                                                       <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;abs_distance&quot;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> \
		                                                                       <span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;geneName&#39;</span><span class="p">,</span> <span class="s1">&#39;motifName&#39;</span><span class="p">],</span> <span class="n">keep</span><span class="o">=</span><span class="s2">&quot;first&quot;</span><span class="p">)</span>

	<span class="n">motifs_genes_and_expression_levels</span><span class="p">[</span><span class="s1">&#39;motifScore&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">motifs_genes_and_expression_levels</span><span class="p">[</span><span class="s1">&#39;motifScore&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>

	<span class="n">TFs_and_associated_expression_profiles</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">motifs_genes_and_expression_levels</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;motifName&#39;</span><span class="p">))</span>
	<span class="n">imputed_TF_features</span> <span class="o">=</span> <span class="p">[]</span>

	<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Performing linear regression for &quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">TFs_and_associated_expression_profiles</span><span class="p">))</span><span class="o">+</span><span class="s2">&quot; transcription factor expression profiles...&quot;</span><span class="p">)</span>

	<span class="k">for</span> <span class="n">TF_name</span><span class="p">,</span> <span class="n">expression_profile</span> <span class="ow">in</span> <span class="n">TFs_and_associated_expression_profiles</span><span class="p">:</span>

		<span class="c1"># Occasionally there&#39;s only one gene associated with a TF, which we can&#39;t fit a line to.</span>
		<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">expression_profile</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">5</span><span class="p">:</span> <span class="k">continue</span>

		<span class="c1"># This allows heavier points to be visualized on the top instead of being hidden by long distance points</span>
		<span class="n">expression_profile</span> <span class="o">=</span> <span class="n">expression_profile</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">expression_profile</span><span class="o">.</span><span class="n">motif_gene_distance</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>

		<span class="c1"># Ordinary Least Squares linear regression</span>
		<span class="n">result</span> <span class="o">=</span> <span class="n">linear_regression</span><span class="p">(</span><span class="n">formula</span><span class="o">=</span><span class="s2">&quot;expression ~ motifScore&quot;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">expression_profile</span><span class="p">)</span><span class="o">.</span><span class="n">fit</span><span class="p">()</span>

		<span class="k">if</span> <span class="n">output_dir</span><span class="p">:</span>
			<span class="n">plot</span> <span class="o">=</span> <span class="n">plot_regression</span><span class="p">(</span><span class="n">model_results</span><span class="o">=</span><span class="n">result</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">expression_profile</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">&quot;motifScore&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;expression&quot;</span><span class="p">,</span> <span class="n">kind</span><span class="o">=</span><span class="s2">&quot;scatter&quot;</span><span class="p">,</span> <span class="n">grid</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>

			<span class="c1"># Add color to points based on distance to gene</span>
			<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">expression_profile</span><span class="p">[</span><span class="s2">&quot;motifScore&quot;</span><span class="p">],</span> <span class="n">expression_profile</span><span class="p">[</span><span class="s2">&quot;expression&quot;</span><span class="p">],</span>
				<span class="n">c</span><span class="o">=</span><span class="p">[</span><span class="nb">abs</span><span class="p">(</span><span class="n">v</span><span class="p">)</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">expression_profile</span><span class="p">[</span><span class="s2">&quot;motif_gene_distance&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()],</span>
				<span class="n">norm</span><span class="o">=</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">colors</span><span class="o">.</span><span class="n">LogNorm</span><span class="p">(</span><span class="n">vmin</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">100000</span><span class="p">,</span> <span class="n">clip</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
				<span class="n">cmap</span><span class="o">=</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">Blues_r</span><span class="p">)</span>
			<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">, </span><span class="si">%0.4f</span><span class="s2">&quot;</span> <span class="o">%</span><span class="p">(</span><span class="n">TF_name</span><span class="p">,</span> <span class="n">result</span><span class="o">.</span><span class="n">pvalues</span><span class="p">[</span><span class="s1">&#39;motifScore&#39;</span><span class="p">]))</span>
			<span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">()</span>

			<span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="s2">&quot;regression_plots&quot;</span><span class="p">),</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
			<span class="n">plot</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="s2">&quot;regression_plots&quot;</span><span class="p">,</span> <span class="n">TF_name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;.png&#39;</span><span class="p">))</span>
			<span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

		<span class="c1"># TODO: implement FDR calculation</span>
		<span class="n">imputed_TF_features</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">TF_name</span><span class="p">,</span> <span class="n">result</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;motifScore&#39;</span><span class="p">],</span> <span class="n">result</span><span class="o">.</span><span class="n">pvalues</span><span class="p">[</span><span class="s1">&#39;motifScore&#39;</span><span class="p">],</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">expression_profile</span><span class="p">[</span><span class="s1">&#39;geneName&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">())))</span>

	<span class="n">imputed_TF_features_dataframe</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">imputed_TF_features</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Transcription Factor&quot;</span><span class="p">,</span> <span class="s2">&quot;Slope&quot;</span><span class="p">,</span> <span class="s2">&quot;P-Value&quot;</span><span class="p">,</span> <span class="s2">&quot;Targets&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;P-Value&quot;</span><span class="p">)</span>

	<span class="c1"># If we&#39;re supplied with an output_dir, we&#39;ll put a summary html file in there as well.</span>
	<span class="k">if</span> <span class="n">output_dir</span><span class="p">:</span>
		<span class="n">html_output</span> <span class="o">=</span> <span class="n">templateEnv</span><span class="o">.</span><span class="n">get_template</span><span class="p">(</span><span class="s2">&quot;summary.jinja&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">render</span><span class="p">(</span><span class="n">images_dir</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span><span class="s2">&quot;regression_plots&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">),</span> <span class="n">TFs</span><span class="o">=</span><span class="nb">sorted</span><span class="p">(</span><span class="n">imputed_TF_features</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">2</span><span class="p">]))</span>
		<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span><span class="s2">&quot;summary.html&quot;</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">summary_output_file</span><span class="p">:</span>
			<span class="n">summary_output_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">html_output</span><span class="p">)</span>

	<span class="k">return</span> <span class="n">imputed_TF_features_dataframe</span></div>


<span class="c1">######################################## Contruct Garnet File ########################################</span>

<span class="k">def</span> <span class="nf">tss_from_bed</span><span class="p">(</span><span class="n">bed_file</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Write a BED file defining the transcription start sites (TSS) inferred from the</span>
<span class="sd">	reference gene file. Sort the TSS and remove any duplicate entries.</span>

<span class="sd">	Arguments:</span>
<span class="sd">		bed_file (str): path to the reference gene BED file</span>

<span class="sd">	Returns:</span>
<span class="sd">		str: the path to the output TSS file.</span>
<span class="sd">	&quot;&quot;&quot;</span>

	<span class="n">output_file</span> <span class="o">=</span> <span class="n">bed_file</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.bed&quot;</span><span class="p">,</span> <span class="s2">&quot;.tss.bed&quot;</span><span class="p">)</span>

	<span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">output_file</span><span class="p">):</span>
		<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;  - TSS file seems to already exists at &#39;</span> <span class="o">+</span> <span class="n">output_file</span><span class="p">)</span>
		<span class="k">return</span> <span class="n">output_file</span>

	<span class="c1"># These series of commands generates a BED file of TSS from the known genes file.</span>
	<span class="c1"># The `awk` command identifies whether the gene is transcribed on the forward or</span>
	<span class="c1"># reverse strand. If the gene is on the forward strand, it assumes the TSS is at the</span>
	<span class="c1"># start of the region, and if the gene is on the reverse strand, the TSS is placed at</span>
	<span class="c1"># the end. The resulting TSS file is sorted and uniquified, and written to output_file.</span>

	<span class="c1"># Code taken from https://github.com/arq5x/bedtools-protocols/blob/master/bedtools.md</span>
	<span class="c1"># TODO: To be replaced by pybedtools command, if possible</span>
	<span class="n">bed_to_tss_cmd</span> <span class="o">=</span> <span class="s1">&#39;&#39;&#39; cat </span><span class="si">%s</span><span class="s1"> | awk &#39;BEGIN{OFS=FS=&quot;</span><span class="se">\t</span><span class="s1">&quot;} </span><span class="se">\</span>
<span class="s1">	       { if ($6 == &quot;+&quot;) </span><span class="se">\</span>
<span class="s1">	         { print $1,$2,$2+1,$4,$5,$6 } </span><span class="se">\</span>
<span class="s1">	         else if ($6 == &quot;-&quot;) </span><span class="se">\</span>
<span class="s1">	         { print $1,$3-1,$3,$4,$5,$6 } </span><span class="se">\</span>
<span class="s1">	       }&#39; </span><span class="se">\</span>
<span class="s1">	| sort -k1,1 -k2,2n </span><span class="se">\</span>
<span class="s1">	| uniq </span><span class="se">\</span>
<span class="s1">	&gt; </span><span class="si">%s</span><span class="s1">&#39;&#39;&#39;</span> <span class="o">%</span><span class="p">(</span><span class="n">bed_file</span><span class="p">,</span> <span class="n">output_file</span><span class="p">)</span>

	<span class="n">subprocess</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">bed_to_tss_cmd</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
	<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;  - Wrote TSS file to &#39;</span> <span class="o">+</span> <span class="n">output_file</span><span class="p">)</span>

	<span class="k">return</span> <span class="n">output_file</span>


<div class="viewcode-block" id="construct_garnet_file"><a class="viewcode-back" href="../index.html#garnet.construct_garnet_file">[docs]</a><span class="k">def</span> <span class="nf">construct_garnet_file</span><span class="p">(</span><span class="n">reference_file</span><span class="p">,</span> <span class="n">motif_file_or_files</span><span class="p">,</span> <span class="n">output_file</span><span class="p">,</span> <span class="n">options</span><span class="o">=</span><span class="nb">dict</span><span class="p">()):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Construct the GarNet file by searching for any motifs that are within</span>
<span class="sd">	a certain window of a reference TSS. Generate a dataframe with these assocations,</span>
<span class="sd">	as well as the distance between the motif and gene TSS. Write the dataframe</span>
<span class="sd">	to the specified output file.</span>

<span class="sd">	Arguments:</span>
<span class="sd">		reference_file (str): path to the reference gene BED file</span>
<span class="sd">		motifs_file (str): path to the motifs BED file</span>
<span class="sd">		output_file (str): ouput GarNet file path</span>
<span class="sd">		options (dict): currently not used</span>

<span class="sd">	Returns:</span>
<span class="sd">		pd.DataFrame: motif-gene associations and distance between the two.</span>
<span class="sd">	&quot;&quot;&quot;</span>

	<span class="c1"># Check whether motif_file_or_files is single path or list of paths</span>
	<span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">motif_file_or_files</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span> <span class="n">motif_files</span> <span class="o">=</span> <span class="n">motif_file_or_files</span>
	<span class="k">else</span><span class="p">:</span> <span class="n">motif_files</span> <span class="o">=</span> <span class="p">[</span><span class="n">motif_file_or_files</span><span class="p">]</span>
	<span class="k">assert</span> <span class="nb">all</span><span class="p">([</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">motif_file</span><span class="p">)</span> <span class="k">for</span> <span class="n">motif_file</span> <span class="ow">in</span> <span class="n">motif_files</span><span class="p">])</span>

	<span class="c1"># Generate BED file of TSS from reference gene file</span>
	<span class="n">reference_tss_file</span> <span class="o">=</span> <span class="n">tss_from_bed</span><span class="p">(</span><span class="n">reference_file</span><span class="p">)</span>
	<span class="n">reference_tss</span> <span class="o">=</span> <span class="n">BedTool</span><span class="p">(</span><span class="n">reference_tss_file</span><span class="p">)</span>

	<span class="c1"># Function that searches for all motifs within a window of 10kb from any gene in the</span>
	<span class="c1"># reference TSS file, and returns a dataframe.</span>
	<span class="c1"># TODO: window size should be generated via the options parameter.</span>
	<span class="n">get_motif_genes_df</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">motif_file</span><span class="p">:</span> <span class="n">reference_tss</span><span class="o">.</span><span class="n">window</span><span class="p">(</span><span class="n">motif_file</span><span class="p">,</span> <span class="n">w</span><span class="o">=</span><span class="mi">10000</span><span class="p">)</span> \
	                                                     <span class="o">.</span><span class="n">to_dataframe</span><span class="p">(</span><span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;tssChrom&quot;</span><span class="p">,</span> <span class="s2">&quot;tssStart&quot;</span><span class="p">,</span> <span class="s2">&quot;tssEnd&quot;</span><span class="p">,</span> <span class="s2">&quot;geneName&quot;</span><span class="p">,</span> <span class="s2">&quot;tssScore&quot;</span><span class="p">,</span> <span class="s2">&quot;tssStrand&quot;</span><span class="p">,</span>
	                                                                          <span class="s2">&quot;motifChrom&quot;</span><span class="p">,</span> <span class="s2">&quot;motifStart&quot;</span><span class="p">,</span> <span class="s2">&quot;motifEnd&quot;</span><span class="p">,</span> <span class="s2">&quot;motifName&quot;</span><span class="p">,</span> <span class="s2">&quot;motifScore&quot;</span><span class="p">,</span> <span class="s2">&quot;motifStrand&quot;</span><span class="p">])</span>

	<span class="c1"># Perform motif-gene matching for each motif file, the concatenate them together.</span>
	<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;  - Searching for motifs near genes. This may take a while...&#39;</span><span class="p">)</span>
	<span class="n">motif_genes_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">get_motif_genes_df</span><span class="p">(</span><span class="n">motif_file</span><span class="p">)</span> <span class="k">for</span> <span class="n">motif_file</span> <span class="ow">in</span> <span class="n">motif_files</span><span class="p">])</span>


	<span class="c1"># Calculate distance between motif and gene by first identifying the side of the motif</span>
	<span class="c1"># that is closest to the gene (this depends on which strand the motif is on), and next</span>
	<span class="c1"># finding the distance between the closest end to the gene TSS. A negative value indicates</span>
	<span class="c1"># that the motif is upstream of the TSS.</span>
	<span class="n">motif_genes_df</span><span class="p">[</span><span class="s2">&quot;motifClosestEnd&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">motif_genes_df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;motifEnd&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;motifStrand&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span> <span class="k">else</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;motifStart&quot;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
	<span class="n">motif_genes_df</span><span class="p">[</span><span class="s2">&quot;motif_gene_distance&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">motif_genes_df</span><span class="p">[</span><span class="s2">&quot;motifClosestEnd&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">motif_genes_df</span><span class="p">[</span><span class="s2">&quot;tssStart&quot;</span><span class="p">])</span> <span class="o">*</span> \
	                                         <span class="n">motif_genes_df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="mi">1</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;tssStrand&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span> <span class="k">else</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

	<span class="c1"># Filter and reorder columns</span>
	<span class="n">motif_genes_df</span> <span class="o">=</span> <span class="n">motif_genes_df</span><span class="p">[[</span><span class="s2">&quot;motifChrom&quot;</span><span class="p">,</span> <span class="s2">&quot;motifStart&quot;</span><span class="p">,</span> <span class="s2">&quot;motifEnd&quot;</span><span class="p">,</span> <span class="s2">&quot;motifName&quot;</span><span class="p">,</span> <span class="s2">&quot;motifScore&quot;</span><span class="p">,</span> <span class="s2">&quot;motifStrand&quot;</span><span class="p">,</span>
	                                 <span class="s2">&quot;geneName&quot;</span><span class="p">,</span> <span class="s2">&quot;tssStart&quot;</span><span class="p">,</span> <span class="s2">&quot;tssEnd&quot;</span><span class="p">,</span> <span class="s2">&quot;motif_gene_distance&quot;</span><span class="p">]]</span>

	<span class="n">motif_genes_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
	<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;  - </span><span class="si">%d</span><span class="s1"> motif-gene associations found and written to </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span><span class="p">(</span><span class="n">motif_genes_df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">output_file</span><span class="p">))</span>

	<span class="k">return</span> <span class="n">motif_genes_df</span></div>

</pre></div>

           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Alexander LeNail.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'2.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>